File: pythonbot_1.0/Player.py
import argparse
import socket
import sys
from HandRankings import hand_win_odds

import PreflopLogic as pl
import FlopTurnLogic as ft
import TurnRiverLogic as tr
import PostRiverLogic as pf

"""
Simple example pokerbot, written in python.

This is an example of a bare bones pokerbot. It only sets up the socket
necessary to connect with the engine and then always returns the same action.
It is meant as an example of how a pokerbot should communicate with the engine.
"""
class Player:


    def run(self, input_socket):
        # Get a file-object for reading packets from the socket.
        # Using this ensures that you get exactly one packet per read.
        f_in = input_socket.makefile()

        while True:
            # Block until the engine sends us a packet.
            data = f_in.readline().strip()
            # If data is None, connection has closed.
            if not data:
                print "Gameover, engine disconnected."
                break

            # Here is where you should implement code to parse the packets from
            # the engine and act on it. We are just printing it instead.

            data_list = data.split()
            word = data_list[0]

            if word == "NEWGAME":
                ourName = data_list[1]

            if word == "NEWHAND":
                hand = [data_list[3], data_list[4]]
                odds = hand_win_odds(hand)
                button = data_list[2]
                streetDict = {'0':0, '3':0, '4':0, '5':0} ##TODO - ADDED DICTIONARY FOR 3-BET LIMITING
                didDiscard = False



            # When appropriate, reply to the engine with a legal action.
            # The engine will ignore all spurious responses.
            # The engine will also check/fold for you if you return an
            # illegal action.
            # When sending responses, terminate each response with a newline
            # character (\n) or your bot will hang!
            if word == "GETACTION":
                action = None
                potsize = int(data_list[1])
                numBoardCards = int(data_list[2])
                boardCards = []  # list of cards on board; len 0, 3, 4, 5
                for i in range(numBoardCards):
                    boardCards.append(data_list[3 + i])

                index = 3 + numBoardCards
                numLastActions = int(data_list[index])
                lastActions = []  # list of previous actions in this hand
                for i in range(numLastActions):
                    lastActions.append(data_list[index + 1 + i])

                index2 = index + 1 + numLastActions
                numMoves = int(data_list[index2])
                moves = []  # list of legal moves
                for i in range(numMoves):
                    moves.append(data_list[index2 + 1 + i])

        #RaiseCounter Variable:
                if str(numBoardCards) in streetDict:
                    streetDict[str(numBoardCards)] += 1



                # At start of Flop, return DISCARD if current hand is not aggressive
                if (streetDict['3'] == 1 and streetDict['4'] == 0) or (streetDict['4'] == 1 and streetDict['5'] == 0):
                    if odds <= 0.55:  # WE ARE GOING TO DISCARD
    #TODO Need to decide WHICH card to discard. Currently discarding low card. Should later take into account boardCards.
                        didDiscard = True
                        cardsDict = {'1': 1, '2': 2, '3': 3, '4': 4, '5': 5, '6': 6, '7': 7, '8': 8, '9': 9, 'T': 10,
                                     'J': 11, 'Q': 12, 'K': 13, 'A': 14}
                        val1 = cardsDict[hand[0][0]]
                        val2 = cardsDict[hand[1][0]]
                        if val1 <= val2:
                            action = ("DISCARD:" + hand[0] + "\n")
                        else:
                            action = ("DISCARD:" + hand[1] + "\n")

                # if didDiscard:          #Makes sure we only change our holeCards if WE were the one to just discard
                for pastAction in lastActions:
                    if "DISCARD" in pastAction and ourName in pastAction:
                        print("UGH MAMI")
                        Old_card_discard = pastAction[8:10]  # Parse old and new card result from Discard action
                        New_card = pastAction[11:13]
                        hand[hand.index(Old_card_discard)] = New_card  # Edit our holeCards list

                        # 1.) figure out current hand from holeCards[]
                        # 2.) parse DISCARD pastAction line (DISCARD:3h:6h:PLAYER) to extract variables Old_card_discard and New_card
                        # 3.) find new hand
                        # 4.) odds = hand_win_odds([Old_card_keep, New_card]) (THIS HAPPENS INSIDE THE LOGIC FILES)
                        break
                        # didDiscard = False

                # After discard, if current hand is now aggressive, turn aggressive to True.



                #decides what LOGIC to follow based on numBoardCards
                if action is None:
                    if numBoardCards == 0:
                        action = pl.getaction(streetDict, potsize, numBoardCards, boardCards, numLastActions, lastActions, numMoves, moves, hand, button)
                    elif numBoardCards == 3:
                        action = ft.getaction(streetDict, potsize, numBoardCards, boardCards, numLastActions, lastActions, numMoves, moves, hand, button)
                    elif numBoardCards == 4:
                        action = tr.getaction(streetDict, potsize, numBoardCards, boardCards, numLastActions, lastActions, numMoves, moves, hand, button)
                    else:
                        action = pf.getaction(streetDict, potsize, numBoardCards, boardCards, numLastActions, lastActions, numMoves, moves, hand, button)
                #print(str(hand) + ", " + str(odds))
                s.send(action + "\n")
            elif word == "REQUESTKEYVALUES":
                # At the end, the engine will allow your bot save key/value pairs.
                # Send FINISH to indicate you're done.
                s.send("FINISH\n")
        # Clean up the socket.
        s.close()



if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='A Pokerbot.', add_help=False, prog='pokerbot')
    parser.add_argument('-h', dest='host', type=str, default='localhost', help='Host to connect to, defaults to localhost')
    parser.add_argument('port', metavar='PORT', type=int, help='Port on host to connect to')
    args = parser.parse_args()

    # Create a socket connection to the engine.
    print 'Connecting to %s:%d' % (args.host, args.port)
    try:
        s = socket.create_connection((args.host, args.port))
    except socket.error as e:
        print 'Error connecting! Aborting'
        exit()

    bot = Player()
    bot.run(s)

File: cppBot/player.cpp
#include <iostream>
#include "player.hpp"

Player::Player() {
}

void Player::run(tcp::iostream &stream) {
    std::string line;
    
    std::string new_game("NEWGAME");
    std::string new_hand("NEWHAND");
    std::string get_action("GETACTION");
    std::string hand_over("HANDOVER");
    std::string request_keyvalue_action("REQUESTKEYVALUES");
    std::string true_string("true");
    std::string first_word;
    std::string action;
    int index_one;
    int index_two;
    
    GameData game_data;
    
    while (std::getline(stream, line)) {
        std::cout << line << "\n";
        
        index_one = line.find_first_of(' ');
        first_word = line.substr(0, index_one));
        if (new_game.compare(first_word) == 0) {
            // extract information from packet
            index_one += 1;
            index_two = line.find(index_one, ' ');
            game_data.name = line.substr(index_one, index_two-index_one);
            index_two += 1;
            index_one = line.find(index_two, ' ');
            game_data.opponent_name = line.substr(index_two, index_one-index_two);
            index_one += 1;
            index_two = line.find(index_one, ' ');
            index_two += 1;
            index_one = line.find(index_two, ' ');
            game_data.big_blind = atoi(line.substr(index_two, index_one-index_two).c_str());
            index_one = line.find_last_of(' ');
            index_one += 1;
            game_data.time_bank = line.substr(index_one, line.length()-index_one);
        } else if (new_hand.compare(first_word) == 0) {
            // extract information from packet
            index_one += 1;
            index_two = line.find(index_one, ' ');
            game_data.hand_id = line.substr(index_one, index_two-index_one);
            index_two += 1;
            index_one = line.find(index_two, ' ');
            game_data.button = true_string.compare(line.substr(index_two, index_one-index_two)) == 0;
            index_one += 1;
            index_two = line.find(index_one, ' ');
            game_data.card_one = line.substr(index_one, index_two-index_one);
            index_two += 1;
            index_one = line.find(index_two, ' ');
            game_data.card_two = line.substr(index_two, index_one-index_two);
            index_one = line.find_last_of(' ');
            index_one += 1;
            game_data.time_bank = line.substr(index_one, line.length()-index_one);
        } else if (get_action.compare(first_word) == 0) {
            // extract information from packet
            index_one += 1;
            index_two = line.find(index_one, ' ');
            game_data.pot_size = atoi(line.substr(index_one, index_two-index_one).c_str());
            index_two += 1;
            index_one = line.find(index_two, ' ');
            int num_board_cards = atoi(line.substr(index_two, index_one-index_two).c_str());
            for (int i = 0; i < num_board_cards; i++) {
                if (i % 2 == 0) {
                    if(game_data.board_cards[i].compare("") == 0) {
                        index_one += 1;
                        index_two = line.find(index_one, ' ');
                        game_data.board_cards[i] = line.substr(index_one, index_two-index_one);
                    }
                } else {
                    if(game_data.board_cards[i].compare("") == 0) {
                        index_two += 1;
                        index_one = line.find(index_two, ' ');
                        game_data.board_cards[i] = line.substr(index_two, index_one-index_two);
                    }
                }
            }
            int num_last_actions;
            if (num_board_cards % 2 == 0) {
                index_one += 1;
                index_two = line.find(index_one, ' ');
                num_last_actions = atoi(line.substr(index_one, index_two-index_one).c_str());
            } else {
                index_two += 1;
                index_one = line.find(index_two, ' ');
                num_last_actions = atoi(line.substr(index_two, index_one-index_two).c_str());
            }
            std::str last_actions[num_last_actions];
            for (int i = 0; i < num_last_actions; i++) {
                if(num_board_cards % 2 != 0) {
                    if (i % 2 == 0) {
                        index_one += 1;
                        index_two = line.find(index_one, ' ');
                        last_actions[i] = line.substr(index_one, index_two-index_one);
	                } else {
                        index_two += 1;
                        index_one = line.find(index_two, ' ');
                        last_actions[i] = line.substr(index_two, index_one-index_two);
	                }
                } else {
                    if (i % 2 != 0) {
                        index_one += 1;
                        index_two = line.find(index_one, ' ');
                        last_actions[i] = line.substr(index_one, index_two-index_one);
                    } else {
                        index_two += 1;
                        index_one = line.find(index_two, ' ');
                        last_actions[i] = line.substr(index_two, index_one-index_two);
                    }
                }
            }
            int num_legal_actions;
            if (num_board_cards % 2 == 0 && num_last_actions % 2 != 0) {
                index_one += 1;
                index_two = line.find(index_one, ' ');
                num_legal_actions = atoi(line.substr(index_one, index_two-index_one).c_str());
            } else {
                index_two += 1;
                index_one = line.find(index_two, ' ');
                num_legal_actions = atoi(line.substr(index_two, index_one-index_two).c_str());
            }
            std::str legal_actions[num_legal_actions];
            for (int i = 0; i < num_legal_actions; i++) {
                if(num_board_cards % 2 != 0 && num_last_actions == 0) {
                    if (i % 2 == 0) {
                        index_one += 1;
                        index_two = line.find(index_one, ' ');
                        legal_actions[i] = line.substr(index_one, index_two-index_one);
                    } else {
                        index_two += 1;
                        index_one = line.find(index_two, ' ');
                        legal_actions[i] = line.substr(index_two, index_one-index_two);
                    }
                } else {
                    if (i % 2 != 0) {
                        index_one += 1;
                        index_two = line.find(index_one, ' ');
                        legal_actions[i] = line.substr(index_one, index_two-index_one);
                    } else {
                        index_two += 1;
                        index_one = line.find(index_two, ' ');
                        legal_actions[i] = line.substr(index_two, index_one-index_two);
                    }
                }
            }
            index_one = line.find_last_of(' ');
            index_one += 1;
            game_data.time_bank = line.substr(index_one, line.length()-index_one);
            std::cout << game_data.pot_size << " " << num_board_cards << " " << num_last_actions << " " << num_legal_actions << " " << game_data.time_bank;
            for (int i = 0; i < num_board_cards; i++) {
                std::cout << game_data.board_cards[i] << endl;
            }
            for (int i = 0; i < num_last_actions; i++) {
                std::cout << last_actions[i] << endl;
            }
            for (int i = 0; i < num_legal_actions; i++) {
                std::cout << legal_actions[i] << endl;
            }
            action = "CHECK";
        } else if (hand_over.compare(first_word) == 0) {
            index_one = line.find_last_of(' ');
            index_one += 1;
            game_data.time_bank = line.substr(index_one, line.length()-index_one);
        } else if (request_keyvalue_action.compare(first_word) == 0) {
            action = "FINISH";
        }
        
        stream << action << "\n";
    }

    std::cout << "Gameover, engine disconnected.\n";
}

struct GameData {
    std::string name;
    std::string opponent_name;
    std::string big_blind;
    std::string time_bank;
    int hand_id;
    bool button;
    int pot_size;
    std::string card_one;
    std::string card_two;
    std::string hand_strength;
    std::string board_cards[5];
};
